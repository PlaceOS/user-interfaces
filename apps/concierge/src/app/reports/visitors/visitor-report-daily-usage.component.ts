import { Component, Input } from '@angular/core';
import { downloadFile, jsonToCsv, unique } from '@placeos/common';
import { map, take } from 'rxjs/operators';
import { VisitorsReportService } from './visitors-report.service';

@Component({
    selector: 'visitor-report-daily-usage',
    template: `
        <div
            class="m-4 rounded bg-base-100 border border-base-200 overflow-hidden"
        >
            <div class="border-b border-base-200 px-4 py-2 flex items-center">
                <h3 class="font-bold text-xl flex-1">Daily Visitor Details</h3>
                <button icon matRipple (click)="download()" *ngIf="!print">
                    <app-icon>download</app-icon>
                </button>
            </div>
            <simple-table
                class="w-full block text-sm"
                [data]="daily_stats"
                [columns]="[
                    { key: 'date', name: 'Date', content: date_template },
                    { key: 'booking_count', name: 'Unique Visitors' },
                    { key: 'host_count', name: 'Host Count' },
                    { key: 'booked_count', name: 'Booking Count' },
                ]"
                [sortable]="true"
                [page_size]="print ? 0 : 10"
                empty_message="No events for selected period"
            ></simple-table>
            <ng-template #date_template let-row="row">
                <div class="p-4">
                    {{ row.date | date: 'mediumDate' }}
                </div>
            </ng-template>
        </div>
    `,
    styles: [``],
})
export class VisitorReportDailyUsageComponent {
    @Input() public print: boolean = false;

    public readonly daily_stats = this._state.daily_stats$.pipe(
        map((days) => {
            let list = [];
            for (const date in days) {
                list.push({
                    date,
                    booking_count: unique(days[date].bookings, 'asset_id')
                        .length,
                    host_count: unique(days[date].bookings, 'user_email')
                        .length,
                    booked_count: days[date].bookings.length,
                });
            }
            return list;
        }),
    );

    public readonly download = async () => {
        const data = await this.daily_stats.pipe(take(1)).toPromise();
        downloadFile('report-visitors-daily-usage.csv', jsonToCsv(data));
    };

    constructor(private _state: VisitorsReportService) {}
}
