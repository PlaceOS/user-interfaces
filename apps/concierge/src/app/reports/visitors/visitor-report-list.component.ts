import { Component, Input } from '@angular/core';
import { downloadFile, jsonToCsv } from '@placeos/common';
import { map, take } from 'rxjs/operators';
import { VisitorsReportService } from './visitors-report.service';

@Component({
    selector: 'visitor-report-list',
    template: `
        <div
            class="m-4 rounded bg-base-100 border border-base-200 overflow-hidden"
        >
            <div class="border-b border-base-200 px-4 py-2 flex items-center">
                <h3 class="font-bold text-xl flex-1">Visitor List</h3>
                <button icon matRipple (click)="download()" *ngIf="!print">
                    <app-icon>download</app-icon>
                </button>
            </div>
            <simple-table
                class="w-full block text-sm"
                [data]="visitor_bookings"
                [columns]="[
                    { key: 'visitor_name', name: 'Visitor Name' },
                    { key: 'date', name: 'Date', content: date_template },
                    { key: 'host', name: 'Host' },
                ]"
                [sortable]="true"
                [page_size]="print ? 0 : 10"
                empty_message="No events for selected period"
            ></simple-table>
            <ng-template #date_template let-row="row">
                <div class="p-4">
                    {{ row.date | date: 'mediumDate' }}
                </div>
            </ng-template>
        </div>
    `,
    styles: [``],
})
export class VisitorReportListComponent {
    @Input() public print: boolean = false;

    public readonly visitor_bookings = this._state.bookings$.pipe(
        map((bookings) => {
            let list = [];
            for (const booking of bookings) {
                list.push({
                    visitor_name:
                        booking.asset_name ||
                        booking.extension_data?.asset_name ||
                        booking.description ||
                        booking.asset_id,
                    date: booking.date,
                    host: booking.user_name || booking.user_email,
                });
            }
            return list;
        }),
    );

    public readonly download = async () => {
        const data = await this.visitor_bookings.pipe(take(1)).toPromise();
        downloadFile('report-assets-daily-usage.csv', jsonToCsv(data));
    };

    constructor(private _state: VisitorsReportService) {}
}
