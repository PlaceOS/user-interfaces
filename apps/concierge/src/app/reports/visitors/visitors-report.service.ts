import { Injectable } from '@angular/core';
import { queryBookings } from '@placeos/bookings';
import {
    downloadFile,
    jsonToCsv,
    notifyError,
    SettingsService,
} from '@placeos/common';
import { OrganisationService } from '@placeos/organisation';
import { endOfDay, format, getUnixTime, isSameDay, startOfDay } from 'date-fns';
import { BehaviorSubject, combineLatest } from 'rxjs';
import {
    debounceTime,
    filter,
    map,
    shareReplay,
    startWith,
    switchMap,
    take,
    tap,
} from 'rxjs/operators';

export interface ReportOptions {
    /** Zones to check available space for */
    zones?: string[];
    /** Start time of the report in ms */
    start?: number;
    /** End time of the report in ms */
    end?: number;
}

@Injectable({
    providedIn: 'root',
})
export class VisitorsReportService {
    private _loading = new BehaviorSubject<boolean>(false);
    private _options = new BehaviorSubject<ReportOptions>({});
    private _generate = new BehaviorSubject<number>(0);

    public readonly loading$ = this._loading.asObservable();
    public readonly options$ = this._options.asObservable();

    public readonly bookings$ = this._generate.pipe(
        filter((gen) => gen > 0),
        switchMap(() => this._options.pipe(take(1))),
        debounceTime(300),
        switchMap((options) => {
            this._loading.next(true);
            const { start, end, zones } = options;
            return queryBookings({
                period_start: getUnixTime(startOfDay(start || Date.now())),
                period_end: getUnixTime(endOfDay(end || start || Date.now())),
                type: 'visitor',
                zones:
                    (zones || [])?.join(',') ||
                    (this._settings.get('app.use_region')
                        ? this._org.region?.id
                        : '') ||
                    this._org.building?.id,
            });
        }),
        tap((_) => {
            if (!_.length) {
                notifyError('No bookings for the selected levels and period');
            }
            this._loading.next(false);
        }),
        startWith([]),
        shareReplay(1),
    );

    public readonly daily_stats$ = combineLatest([
        this.bookings$,
        this._options,
    ]).pipe(
        map(([bookings, options]) => {
            const days = {};
            for (const booking of bookings) {
                const date = format(booking.date, 'yyyy-MM-dd');
                if (!days[date]) {
                    days[date] = {
                        date: booking.date,
                        bookings: [],
                    };
                }
                days[date].bookings.push(booking);
            }
            return days;
        }),
        shareReplay(1),
    );

    constructor(
        private _settings: SettingsService,
        private _org: OrganisationService,
    ) {}

    public setOptions(options: Partial<ReportOptions>) {
        this._options.next({ ...this._options.getValue(), ...options });
    }

    public generateReport() {
        this._generate.next(Date.now());
    }

    public async downloadReport() {
        const options = this._options.getValue();
        const bookings = await this.bookings$.pipe(take(1)).toPromise();
        if (!bookings?.length) return;
        const is_same = isSameDay(options.start, options.end);
        const date = is_same
            ? format(options.start, 'yyyy-MM-dd')
            : `${format(options.start, 'yyyy-MM-dd')}-${format(
                  options.end,
                  'yyyy-MM-dd',
              )}`;
        downloadFile(
            `report+assets+${date}.tsv`,
            jsonToCsv(
                bookings.map((bkn: any) => {
                    const details = bkn.toJSON();
                    delete details.zones;
                    delete details.server_names;
                    delete details.extension_data;
                    return details;
                }),
                '\t',
            ),
        );
    }
}
