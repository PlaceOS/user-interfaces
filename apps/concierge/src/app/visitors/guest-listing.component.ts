import { Component } from '@angular/core';
import { AsyncHandler, SettingsService, notifyError } from '@placeos/common';
import { VisitorsStateService } from './visitors-state.service';
import { Booking } from '@placeos/bookings';
import { showMetadata } from '@placeos/ts-client';
import { OrganisationService } from '@placeos/organisation';

@Component({
    selector: 'guest-listings',
    template: `
        <simple-table
            class="min-w-[60rem] block text-sm z-0"
            [data]="guests"
            [columns]="[
                {
                    key: 'state',
                    name: 'Checked In',
                    content: state_template,
                    size: '6.5rem',
                    sortable: false
                },
                {
                    key: 'date',
                    name: 'Time',
                    content: date_template,
                    size: '6rem'
                },
                {
                    key: 'asset_name',
                    name: 'Visitor',
                    content: person_template
                },
                { key: 'user_name', name: 'Host', content: host_template },
                {
                    key: 'status',
                    name: 'State',
                    content: status_template,
                    size: '9.5rem'
                },
                {
                    key: 'induction',
                    name: 'Inducted',
                    content: boolean_template,
                    show: !!inductions_enabled,
                    size: '5.5rem'
                },
                {
                    key: 'actions',
                    name: ' ',
                    content: action_template,
                    size: '3.25rem',
                    sortable: false
                }
            ]"
            [filter]="search | async"
            [sortable]="true"
        ></simple-table>
        <ng-template #state_template let-row="row">
            <div
                *ngIf="!row?.checked_in && row.checked_out_at"
                class="rounded h-8 w-8 flex items-center justify-center text-2xl bg-base-400 text-neutral-content mx-auto"
                [matTooltip]="
                    'Checked out at ' + (row.checked_out_at | date: time_format)
                "
                matTooltipPosition="right"
            >
                <app-icon>done</app-icon>
            </div>
            <div
                *ngIf="!row?.checked_in && !row.checked_out_at"
                class="rounded h-8 w-8 flex items-center justify-center text-2xl bg-warning text-warning-content mx-auto"
                matTooltip="Not checked in"
                matTooltipPosition="right"
            >
                <app-icon>question_mark</app-icon>
            </div>
            <div
                *ngIf="row?.checked_in"
                class="rounded h-8 w-8 flex items-center justify-center text-2xl bg-success text-success-content mx-auto"
                matTooltip="Checked In"
                matTooltipPosition="right"
            >
                <app-icon>done</app-icon>
            </div>
        </ng-template>
        <ng-template #person_template let-row="row">
            <div class="px-4 py-2">
                <div>{{ row.asset_name || row.asset_id }}</div>
                <div
                    *ngIf="row.asset_name && row.asset_id"
                    class="opacity-30 text-xs"
                >
                    {{ row.asset_id }}
                </div>
            </div>
        </ng-template>
        <ng-template #host_template let-row="row">
            <div class="px-4 py-2">
                <div>{{ row.user_name || row.user_email }}</div>
                <div
                    *ngIf="row.user_name && row.user_email"
                    class="opacity-30 text-xs"
                >
                    {{ row.user_email }}
                </div>
            </div>
        </ng-template>
        <ng-template #id_template let-row="row">
            <div customTooltip [content]="id_confirmation">
                <button
                    matRipple
                    *ngIf="row.extension_data?.id_data?.url"
                    class="bg-success rounded-3xl px-4 py-2 text-white"
                >
                    {{
                        row.extension_data?.id_confirmed
                            ? 'Confirmed'
                            : row.extension_data?.id_confirmed === false
                            ? 'Rejected'
                            : 'Submitted'
                    }}
                </button>
            </div>
            <ng-template #id_confirmation>
                <div
                    class="bg-base-100 rounded p-2 flex flex-col space-y-2 my-2 w-[20rem]"
                >
                    <img
                        [src]="row.extension_data?.id_data?.url"
                        class="max-w-[20rem] max-h-[20rem] object-contain"
                    />
                    <button
                        matRipple
                        (click)="setExt(row, 'id_confirmed', true)"
                    >
                        Confirm ID
                    </button>
                    <button
                        matRipple
                        class="inverse mt-2"
                        (click)="setExt(row, 'id_confirmed', false)"
                    >
                        Reject ID
                    </button>
                </div>
            </ng-template>
        </ng-template>
        <ng-template #boolean_template let-row="row">
            <div
                *ngIf="inducted(row)"
                class="rounded h-8 w-8 flex items-center justify-center text-2xl bg-success text-success-content mx-auto"
            >
                <app-icon>done</app-icon>
            </div>
            <div
                *ngIf="inducted(row) === null"
                class="rounded h-8 w-8 flex items-center justify-center text-2xl bg-warning text-warning-content mx-auto"
            >
                <app-icon>question_mark</app-icon>
            </div>
            <div
                *ngIf="inducted(row) === false"
                class="rounded h-8 w-8 flex items-center justify-center text-2xl bg-error text-error-content mx-auto"
            >
                <app-icon>close</app-icon>
            </div>
        </ng-template>
        <ng-template #status_template let-row="row">
            <div class="px-4">
                <button
                    matRipple
                    class="rounded-3xl bg-warning text-warning-content border-none w-[7.5rem] h-10"
                    [class.!text-success-content]="row?.status === 'approved'"
                    [class.!bg-success]="row?.status === 'approved'"
                    [class.!text-error-content]="row?.status === 'declined'"
                    [class.!bg-error]="row?.status === 'declined'"
                    [class.!text-neutral-content]="row?.status === 'ended'"
                    [class.!bg-neutral]="row?.status === 'ended'"
                    [class.opacity-30]="row?.status === 'ended'"
                    [matMenuTriggerFor]="menu"
                    [disabled]="
                        row?.status === 'ended' ||
                        (row.checked_in && !row.checked_out_at)
                    "
                >
                    <div class="flex items-center pl-4 pr-2 space-x-2">
                        <div class="flex-1 text-left">
                            {{
                                row?.status === 'ended'
                                    ? 'Ended'
                                    : row?.status === 'approved'
                                    ? 'Approved'
                                    : row?.status === 'declined'
                                    ? 'Declined'
                                    : 'Pending'
                            }}
                        </div>
                        <app-icon
                            class="text-2xl"
                            *ngIf="
                                !(
                                    row?.status === 'ended' ||
                                    (row.checked_in && !row.checked_out_at)
                                )
                            "
                        >
                            arrow_drop_down
                        </app-icon>
                    </div>
                </button>
            </div>
            <mat-menu #menu="matMenu">
                <button mat-menu-item (click)="approveVisitor(row)">
                    <div class="flex items-center space-x-2">
                        <app-icon class="text-2xl">event_available</app-icon>
                        <div class="pr-2">Approve Visitor</div>
                    </div>
                </button>
                <button mat-menu-item (click)="declineVisitor(row)">
                    <div class="flex items-center space-x-2">
                        <app-icon class="text-2xl text-error">
                            event_busy
                        </app-icon>
                        <div class="pr-2">Decline Visitor</div>
                    </div>
                </button>
            </mat-menu>
        </ng-template>
        <ng-template #date_template let-row="row">
            <div class="px-4">
                {{
                    row.date
                        | date
                            : ((filters | async)?.period > 1
                                  ? 'MMM d, ' + time_format
                                  : time_format)
                }}
            </div>
        </ng-template>
        <ng-template #action_template let-row="row">
            <div class="flex items-center justify-end px-2">
                <button icon matRipple [matMenuTriggerFor]="guest_menu">
                    <app-icon>more_horiz</app-icon>
                </button>
                <mat-menu #guest_menu="matMenu">
                    <button
                        mat-menu-item
                        [disabled]="!row.attachment?.length"
                        [matMenuTriggerFor]="attachment_menu"
                    >
                        <div class="flex items-center space-x-2">
                            <app-icon class="text-2xl">attachment</app-icon>
                            <div>View Attachments</div>
                        </div>
                    </button>
                    <mat-menu #menu="matMenu">
                        <a
                            *ngFor="let item of row.attachments"
                            [href]="item.url"
                            mat-menu-item
                        >
                            {{ item.name }}
                        </a>
                    </mat-menu>
                    <button mat-menu-item (click)="setExt(row, 'remote', true)">
                        <div class="flex items-center space-x-2">
                            <app-icon class="text-2xl">
                                {{
                                    row.extension_data.remote
                                        ? 'business'
                                        : 'laptop'
                                }}
                            </app-icon>
                            <div>
                                Set as
                                {{
                                    row.extension_data.remote
                                        ? 'Onsite'
                                        : 'Remote'
                                }}
                                Visitor
                            </div>
                        </div>
                    </button>
                    <button
                        mat-menu-item
                        *ngIf="can_print"
                        (click)="printQRCode()"
                    >
                        <div class="flex items-center space-x-2">
                            <app-icon class="text-2xl">print</app-icon>
                            <div>Print QR Code</div>
                        </div>
                    </button>
                    <a mat-menu-item [href]="'mailto:' + row?.asset_id">
                        <div class="flex items-center space-x-2">
                            <app-icon class="text-2xl">email</app-icon>
                            <div>
                                Email
                                {{
                                    row?.user_email === row?.asset_id
                                        ? 'Host'
                                        : 'Guest'
                                }}
                            </div>
                        </div>
                    </a>
                    <button
                        mat-menu-item
                        (click)="row.checked_in ? checkout(row) : checkin(row)"
                        *ngIf="!row.checked_out_at"
                    >
                        <div class="flex items-center space-x-2">
                            <app-icon class="text-2xl">
                                {{
                                    row.checked_in
                                        ? 'event_busy'
                                        : 'event_available'
                                }}
                            </app-icon>
                            <div>
                                {{ row.checked_in ? 'Checkout' : 'Checkin' }}
                                Guest
                            </div>
                        </div>
                    </button>
                    <ng-container *ngIf="row.linked_event">
                        <button mat-menu-item (click)="checkinAllVisitors(row)">
                            <div class="flex items-center space-x-2">
                                <app-icon class="text-2xl">
                                    event_available
                                </app-icon>
                                <div>Checkin all for Meeting</div>
                            </div>
                        </button>
                        <button
                            mat-menu-item
                            (click)="checkoutAllVisitors(row)"
                        >
                            <div class="flex items-center space-x-2">
                                <app-icon class="text-2xl text-error">
                                    event_busy
                                </app-icon>
                                <div>Checkout all for Meeting</div>
                            </div>
                        </button>
                    </ng-container>
                </mat-menu>
            </div>
        </ng-template>
        <button
            class="bg-secondary hover:shadow-lg shadow absolute bottom-4 right-4 text-white h-12 w-12 z-20"
            matTooltip="Download Visitor List"
            matTooltipPosition="left"
            icon
            matRipple
            *ngIf="(guests | async)?.length"
            (click)="downloadVisitorList()"
        >
            <app-icon>download</app-icon>
        </button>
        <div class="w-full h-8"></div>
    `,
    styles: [``],
})
export class GuestListingComponent extends AsyncHandler {
    public readonly guests = this._state.filtered_bookings;
    public readonly search = this._state.search;
    public readonly filters = this._state.filters;
    public inductions_enabled = false;

    public readonly downloadVisitorList = () =>
        this._state.downloadVisitorsList();

    public readonly approveVisitor = (u) => this._state.approveVisitor(u);
    public readonly declineVisitor = (u) => this._state.declineVisitor(u);
    public readonly checkinAllVisitors = (u) =>
        this._state.setCheckinStateForEvent(u.linked_event?.id, true);
    public readonly checkoutAllVisitors = (u) =>
        this._state.setCheckinStateForEvent(u.linked_event?.id, false);
    public readonly setExt = (u, f, v) => this._state.setExt(u, f, v);

    public readonly checkin = async (item: Booking) => {
        await this._state.setCheckinState(item, true).catch((e) => {
            if (e !== 'User declined') notifyError(e);
            throw e;
        });
        this._state.poll();
    };

    public readonly checkout = async (item: Booking) => {
        await this._state.setCheckinState(item, false);
        this._state.poll();
    };

    public get time_format() {
        return this._settings.time_format;
    }

    public inducted(item: Booking) {
        if (item.checked_in) return true;
        return item.process_state.includes('declined')
            ? false
            : item.process_state.includes('inducted') || item.induction
            ? true
            : null;
    }

    constructor(
        private _state: VisitorsStateService,
        private _settings: SettingsService,
        private _org: OrganisationService
    ) {
        super();
    }

    public ngOnInit() {
        this.subscription(
            'building',
            this._org.active_building.subscribe(async (bld) => {
                if (!bld) return;
                const visitor_kiosk_app =
                    this._settings.get('app.visitor_kiosk_app') ||
                    'visitor-kiosk_app';
                const metadata: any = await showMetadata(
                    bld.id,
                    visitor_kiosk_app
                ).toPromise();
                this.inductions_enabled =
                    metadata.details?.induction_enabled &&
                    metadata.details?.induction_details;
            })
        );
    }
}
