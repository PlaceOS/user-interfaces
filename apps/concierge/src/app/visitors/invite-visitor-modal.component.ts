import { Component } from '@angular/core';
import { MatDialogRef } from '@angular/material/dialog';

@Component({
    selector: 'invite-visitor-modal',
    template: `
        <main class="relative min-w-[40rem] min-h-[40rem]">
            <invite-visitor-form (done)="onDone()"></invite-visitor-form>
        </main>
    `,
    styles: [``],
})
export class InviteVisitorModalComponent {
    constructor(
        private _dialog_ref: MatDialogRef<InviteVisitorModalComponent>
    ) {}

    public onDone() {
        this._dialog_ref.close();
    }
}
