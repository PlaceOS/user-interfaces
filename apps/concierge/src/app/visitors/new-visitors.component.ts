import { Component, OnDestroy, OnInit } from '@angular/core';
import { VisitorsStateService } from './visitors-state.service';
import { OrganisationService } from '@placeos/organisation';
import { ActivatedRoute, Router } from '@angular/router';
import { MatDialog } from '@angular/material/dialog';
import { InviteVisitorModalComponent } from './invite-visitor-modal.component';
import { combineLatest } from 'rxjs';
import { map } from 'rxjs/operators';
import { SettingsService } from '@placeos/common';

@Component({
    selector: '[app-new-visitors]',
    template: `
        <app-topbar></app-topbar>
        <div class="flex flex-1 h-px">
            <app-sidebar></app-sidebar>
            <main class="flex flex-col flex-1 w-1/2 h-full">
                <div class="flex items-center px-8 py-4 space-x-2">
                    <h2 class="text-2xl font-medium">Visitors</h2>
                    <div class="flex-1 w-2"></div>
                    <searchbar
                        class="mr-2"
                        (modelChange)="setSearch($event)"
                    ></searchbar>
                    <button btn matRipple (click)="inviteVisitor()">
                        Invite Visitor
                    </button>
                </div>
                <div class="flex items-center px-8 pb-4">
                    <mat-form-field
                        appearance="outline"
                        class="no-subscript w-60"
                    >
                        <mat-select
                            [ngModel]="(filters | async)?.zones"
                            (ngModelChange)="updateZones($event)"
                            placeholder="All Levels"
                            multiple
                        >
                            <mat-option
                                *ngFor="let level of levels | async"
                                [value]="level.id"
                            >
                                <div class="flex flex-col-reverse">
                                    <div
                                        class="text-xs opacity-30"
                                        *ngIf="use_region"
                                    >
                                        {{
                                            (level.parent_id | building)
                                                ?.display_name
                                        }}
                                        <span class="opacity-0"> - </span>
                                    </div>
                                    <div>
                                        {{ level.display_name || level.name }}
                                    </div>
                                </div>
                            </mat-option>
                        </mat-select>
                    </mat-form-field>
                    <div class="flex-1 w-2"></div>
                    <date-options (dateChange)="setDate($event)"></date-options>
                </div>
                <div class="mx-8 flex-1 h-1/2 overflow-auto">
                    <guest-listings></guest-listings>
                </div>
                <mat-progress-bar
                    class="w-full"
                    *ngIf="loading | async"
                    mode="indeterminate"
                ></mat-progress-bar>
            </main>
        </div>
    `,
    styles: [
        `
            :host {
                display: flex;
                flex-direction: column;
                height: 100%;
                width: 100%;
                background-color: var(--b1);
            }
        `,
    ],
})
export class NewVisitorsComponent implements OnInit, OnDestroy {
    public readonly loading = this._state.loading;
    public readonly filters = this._state.filters;
    /** List of selected levels */
    public zones: string[] = [];
    /** List of levels for the active building */
    public readonly levels = combineLatest([
        this._org.active_building,
        this._org.active_region,
    ]).pipe(
        map(([bld, region]) =>
            this._settings.get('app.use_region')
                ? this._org.levelsForRegion(region)
                : this._org.levelsForBuilding(bld)
        )
    );
    /** Set filtered date */
    public readonly setDate = (date) => this._state.setFilters({ date });
    /** Set filtered date */
    public readonly setFilters = (filters) => this._state.setFilters(filters);
    /** Set filter string */
    public readonly setSearch = (str) => this._state.setSearchString(str);
    /** Update active zones for desks */
    public readonly updateZones = (zones) => {
        this._router.navigate([], {
            relativeTo: this._route,
            queryParams: { zone_ids: zones.join(',') },
        });
        this._state.setFilters({ zones });
    };

    public get use_region() {
        return !!this._settings.get('app.use_region');
    }

    constructor(
        private _state: VisitorsStateService,
        private _org: OrganisationService,
        private _router: Router,
        private _route: ActivatedRoute,
        private _dialog: MatDialog,
        private _settings: SettingsService
    ) {}

    public inviteVisitor() {
        this._dialog.open(InviteVisitorModalComponent);
    }

    public ngOnInit() {
        this._state.startPolling();
    }

    public ngOnDestroy() {
        this._state.stopPolling();
    }
}
