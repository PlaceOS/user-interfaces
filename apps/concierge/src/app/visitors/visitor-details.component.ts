import {
    Component,
    EventEmitter,
    Input,
    OnChanges,
    Output,
    SimpleChanges,
} from '@angular/core';
import { BaseClass, notifyError, SettingsService } from '@placeos/common';
import { CalendarEvent, saveEvent } from '@placeos/events';
import { showGuest, User } from '@placeos/users';

import { VisitorsStateService } from './visitors-state.service';

@Component({
    selector: 'visitor-details',
    template: `
        <div class="w-12 text-lg flex justify-center">
            <ng-container *ngIf="!visitor?.organizer; else host_state">
                <i
                    *ngIf="!visitor?.checked_in; else checkin_state"
                    class="p-2 rounded-full material-icons border-2 border-dotted border-gray-600"
                    >face</i
                >
            </ng-container>
            <ng-template #checkin_state>
                <i
                    class="p-2 rounded-full material-icons bg-green-600 border-2 border-green-600 text-white"
                    >done</i
                >
            </ng-template>
            <ng-template #host_state>
                <i
                    class="p-2 rounded-full material-icons bg-blue-600 border-2 border-blue-600 text-white"
                    >assignment_ind</i
                >
            </ng-template>
        </div>
        <div flex class="p-2 flex-1">{{ visitor?.name || visitor?.email }}</div>
        <div class="w-48 py-2 flex items-center justify-end">
            <action-icon
                remote
                [matTooltip]="
                    remote
                        ? 'Set as In-Person Visitor'
                        : 'Set as Remote Visitior'
                "
                [loading]="loading === 'remote'"
                [content]="remote ? 'tap_and_play' : 'business'"
                (click)="toggleRemote()"
                [class.invisible]="!visitor?.is_external || visitor?.organizer"
            >
            </action-icon>
            <action-icon
                checkin
                matTooltip="Checkin Guest"
                [loading]="loading === 'checkin'"
                [state]="visitor?.checked_in ? 'success' : ''"
                content="event_available"
                (click)="checkin()"
                [class.invisible]="!visitor?.is_external || visitor?.organizer"
            >
            </action-icon>
            <action-icon
                checkout
                matTooltip="Checkout Guest"
                [loading]="loading === 'checkout'"
                content="event_busy"
                (click)="checkout()"
                [class.invisible]="!visitor?.is_external || visitor?.organizer"
            >
            </action-icon>
            <a
                [href]="'mailto:' + visitor?.email"
                mat-icon-button
                [matTooltip]="visitor?.organizer ? 'Email Host' : 'Email Guest'"
            >
                <app-icon>email</app-icon>
            </a>
            <action-icon
                matTooltip="Print QR Code"
                [loading]="loading === 'printing'"
                content="event_busy"
                (click)="printQRCode()"
                [class.invisible]="!can_print"
            >
            </action-icon>
        </div>
        <div class="w-14 p-2"></div>
        <div
            qr-code
            *ngIf="show_qr_code"
            class="print-only fixed inset-0 flex flex-col justify-center bg-white space-y-2"
        >
            <h2>{{ visitor?.name || visitor?.email }}</h2>
            <div>
                <span>Host:</span>
                {{
                    event?.organiser?.name ||
                        event?.organiser?.name ||
                        event?.creator
                }}
            </div>
            <div><span>Purpose:</span> {{ event?.title }}</div>
            <div>
                <span>Location:</span>
                {{ event?.location || '&lt;Unspecified&gt;' }}
            </div>
            <div><span>Issue Date:</span> {{ today | date: 'mediumDate' }}</div>
            <div class="flex justify-center">
                <img
                    [src]="visitor?.extension_data?.qr?.code || ''"
                    alt="qr-code"
                />
            </div>
        </div>
    `,
    styles: [
        `
            :host {
                display: flex;
                align-items: center;
                width: 100%;
                padding: 0 0.5rem;
            }

            :host > div {
                display: flex;
                align-items: center;
                height: 100%;
                border-bottom: 1px solid #ccc;
                height: 3.5rem;
            }

            :host > div:first-child {
                border: none;
            }

            .invisible {
                opacity: 0;
                pointer-events: none;
            }
        `,
    ],
})
export class VisitorDetailsComponent extends BaseClass implements OnChanges {
    @Input() public event: CalendarEvent;
    @Input() public visitor: User;
    @Output() public eventChange = new EventEmitter<CalendarEvent>();

    public show_attendees: boolean;
    public loading: string;
    public show_qr_code: boolean;

    public readonly checkin = async () => {
        this.loading = 'checkin';
        this.event = await this._state
            .checkGuestIn(this.event, this.visitor)
            .catch((e) => this.event);
        this.eventChange.emit(this.event);
        this.loading = '';
    };

    public readonly toggleRemote = async () => {
        this.loading = 'remote';
        const remote_list =
            this.event.ext('remote')?.filter((e) => e !== this.visitor.email) ||
            [];
        if (!this.remote) {
            remote_list.push(this.visitor.email);
        }
        const new_item = new CalendarEvent({
            ...this.event.toJSON(),
            remote: remote_list,
        }).toJSON();
        this.event = await saveEvent(new_item)
            .toPromise()
            .catch((e) => {
                notifyError(
                    `Error setting visitor status. Error: ${
                        e.statusText || e.message || e
                    }`
                );
                return this.event;
            });
        this.eventChange.emit(this.event);
        this.loading = '';
    };

    public readonly checkout = async () => {
        this.loading = 'checkout';
        this.event = await this._state
            .checkGuestOut(this.event, this.visitor)
            .catch((e) => this.event);
        this.eventChange.emit(this.event);
        this.loading = '';
    };

    /** Open print dialog for user's QR code */
    public readonly printQRCode = () => {
        this.show_qr_code = true;
        this.timeout(
            'print',
            () => {
                window.print();
                this.show_qr_code = false;
            },
            50
        );
    };

    public get can_print(): boolean {
        return (
            this._settings.get('app.can_print_qr') &&
            this.visitor.checked_in &&
            this.visitor?.extension_data.qr
        );
    }

    public get today(): number {
        return Date.now();
    }

    public get remote(): boolean {
        return !!this.event?.extension_data?.remote?.find(
            (e) => e === this.visitor?.email
        );
    }

    constructor(
        private _state: VisitorsStateService,
        private _settings: SettingsService
    ) {
        super();
    }

    public ngOnChanges(changes: SimpleChanges) {
        if (changes.visitor) this.loadGuest();
    }

    public async loadGuest(tries: number = 0) {
        if (!this.visitor) return;
        this.loading = 'printing';
        if (this.visitor.checked_in) {
            const guest = await showGuest(this.visitor.email)
                .toPromise()
                .catch((_) => null);
            if (!guest?.extension_data?.qr?.code && tries < 5) {
                this.timeout('load_guest', () => this.loadGuest(++tries), 1000);
            } else this.loading = '';
        } else this.loading = '';
    }
}
