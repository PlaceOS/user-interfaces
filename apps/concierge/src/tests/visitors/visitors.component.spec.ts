import { MatProgressBarModule } from '@angular/material/progress-bar';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { MockComponent } from 'ng-mocks';
import { BehaviorSubject } from 'rxjs';

import { SidebarComponent } from '../../app/ui/sidebar.component';
import { VisitorListingsComponent } from '../../app/visitors/visitor-listings.component';
import { VisitorsStateService } from '../../app/visitors/visitors-state.service';
import { VisitorsTopbarComponent } from '../../app/visitors/visitors-topbar.component';
import { VisitorsComponent } from '../../app/visitors/visitors.component';

describe('VisitorsComponent', () => {
    let spectator: Spectator<VisitorsComponent>;
    const createComponent = createComponentFactory({
        component: VisitorsComponent,
        providers: [
            {
                provide: VisitorsStateService,
                useValue: {
                    loading: new BehaviorSubject(false),
                    startPolling: jest.fn(),
                    stopPolling: jest.fn(),
                },
            },
        ],
        declarations: [
            MockComponent(SidebarComponent),
            MockComponent(VisitorsTopbarComponent),
            MockComponent(VisitorListingsComponent),
        ],
        imports: [MatProgressBarModule],
    });

    beforeEach(() => (spectator = createComponent()));

    it('should create component', () => {
        expect(spectator.component).toBeTruthy();
    });

    it('should match snapshot', () => {
        expect(spectator.element).toMatchSnapshot();
    });
});
