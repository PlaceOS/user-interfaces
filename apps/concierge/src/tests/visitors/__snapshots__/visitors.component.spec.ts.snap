// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VisitorsComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <sidebar />
  <main
    class="relative overflow-hidden flex-1 flex flex-col"
  >
    <visitors-topbar
      class="w-full"
    />
    <visitor-listings
      class="w-full flex-1 h-0"
    />
    
  </main>
</div>
`;
